crates/shared/src/bad_token/cache.rs

```diff
@@ -94,11 +86,12 @@ impl CachingDetector {
                 let start = Instant::now();
 
                 let expired_tokens: Vec<H160> = {
-                    let cache = detector.cache.read().await;
                     let now = Instant::now();
-                    cache
+                    detector
+                        .cache
                         .iter()
-                        .filter_map(|(token, (instant, _))| {
+                        .filter_map(|entry| {
+                            let (token, (instant, _)) = entry.pair();
                             (now.checked_duration_since(*instant).unwrap_or_default()
                                 >= prefetch_time_to_expire)
```

Comment:
I know that you didn't touch this but having the named variable might still be nicer to read.
```suggestion
                            let valid =
                                now.saturating_duration_since(*instant) >= prefetch_time_to_expire;
                            valid.then_some(*token)
```